ui/app/js/controllers/user_edit_controller.js

```diff
@@ -31,12 +31,96 @@ function ($scope, $modalInstance, CSRF, Permissions, user, users) {
     $scope.loading = false;
   });
 
+  $scope.user.roles = [];
+  Permissions.getUserRoles(user.username)
+  .success(function(response) {
+    $scope.user.roles = response.roles;
+  })
+  .error(function(response) {
+    if (typeof response === 'object') {
+      $scope.errors = response;
+      sweetAlert(
+        "Form submission error",
+        "See fields highlighted in red.",
+        "error"
+      );
+    } else if (typeof response === 'string') {
+      sweetAlert(
+        "Form submission error",
+        "Unable to submit successfully.\n" +
+        "(" + response+ ")",
+        "error"
+      );
+    }
+  });
+
+  $scope.roles_list = [];
+  Permissions.getAllRoles()
+  .success(function(response) {
+    $scope.roles_list = response.roles;
+  })
+  .error(function(response) {
+    if (typeof response === 'object') {
+      $scope.errors = response;
+      sweetAlert(
+        "Form submission error",
+        "See fields highlighted in red.",
+        "error"
+      );
+    } else if (typeof response === 'string') {
+      sweetAlert(
+        "Form submission error",
+        "Unable to submit successfully.\n" +
+        "(" + response+ ")",
+        "error"
+      );
+    }
+  });
+
   $scope.saving = false;
 
   $scope.$watchCollection('permission', function(value) {
     value.options = value.options_as_json;
   });
 
+  $scope.grantRole = function() {
+    $scope.saving = true;
+    CSRF.getToken()
+    .then(function(csrf_token) {
+      Permissions.grantRole($scope.user.username, $scope.role.role, $scope.role.data_version, csrf_token)
+      .success(function(response) {
+        $scope.user.roles.push($scope.role);
```

Comment:
I think you need to parse the response for "new_data_version" and set it in the right place. Right now, if I try to revoke a role immediately after granting it, I get "data version is not a valid integer value".